pkg/cache/inmemory.go

```diff
@@ -0,0 +1,302 @@
+func NewInMemoryCache(name string, logger log.Logger, reg prometheus.Registerer, conf []byte) (*InMemoryCache, error) {
+	config, err := parseInMemoryCacheConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewInMemoryCacheWithConfig(name, logger, reg, config)
+}
+
+// NewInMemoryCacheWithConfig creates a new thread-safe LRU cache chunks and metadata ensures the total cache
+// size approximately does not exceed maxBytes.
+func NewInMemoryCacheWithConfig(name string, logger log.Logger, reg prometheus.Registerer, config InMemoryCacheConfig) (*InMemoryCache, error) {
+	if config.MaxItemSize > config.MaxSize {
+		return nil, errors.Errorf("max item size (%v) cannot be bigger than overall cache size (%v)", config.MaxItemSize, config.MaxSize)
+	}
+
+	c := &InMemoryCache{
+		logger:           logger,
+		maxSizeBytes:     uint64(config.MaxSize),
+		maxItemSizeBytes: uint64(config.MaxItemSize),
+	}
+
+	c.evicted = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_store_cache_items_evicted_total",
+		Help:        "Total number of items that were evicted from the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.added = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_store_cache_items_added_total",
+		Help:        "Total number of items that were added to the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.requests = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_store_cache_requests_total",
+		Help:        "Total number of requests to the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.hitsExpired = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_store_cache_hits_on_expired_data_total",
+		Help:        "Total number of requests to the inmemory cache that were a hit but needed to be evicted due to TTL.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.overflow = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_store_cache_items_overflowed_total",
+		Help:        "Total number of items that could not be added to the inmemory cache due to being too big.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.hits = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name:        "thanos_store_cache_hits_total",
+		Help:        "Total number of requests to the inmemory cache that were a hit.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.current = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name:        "thanos_store_cache_items",
+		Help:        "Current number of items in the inmemory cache.",
+		ConstLabels: prometheus.Labels{"name": name},
+	})
+
+	c.currentSize = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name: "thanos_store_cache_items_size_bytes",
+		Help: "Current byte size of items in the inmemory cache.",
```

Comment:
I think you forgot to add `ConstLabels` here (and in the next ones too).

Reply:
nice catch!